### PR TITLE
Implement ranges::uninitialized_move

### PR DESCRIPTION
**Pull request by CaseyCarter**
`<algorithm>`
* Promote `_Convertible_from` and `in_out_result` to `<xmemory>` for use by the specialized memory algorithms in`<memory>`, (notably including `ranges::uninitialized_move`).

`<xmemory>`
* Teach `_Destroy_in_place` to destroy arrays.
* Teach `_Destroy_range` to accept different sentinel types.

`<memory>`
* Implement exposition-only concepts `_No_throw_input_iterator`, `_No_throw_sentinel_for`, `_No_throw_forward_iterator`, `_No_throw_input_range`, and `_No_throw_forward_range` (with test coverage in `tests/std/tests/P0896R4_ranges_algorithm_machinery`).
* Implement `ranges::uninitialized_move` (with test `tests/std/tests/P0896R4_ranges_alg_uninitialized_move`).

`tests/std/include/range_algorithm_support.hpp`
* `test::iterator`'s `iter_move` should yield an xvalue instead of a prvalue to eliminate the redundant extra move in `*x = iter_move(y)`.

### CODE REVIEW

File: stl/inc/xmemory [hunk 1]
```diff
@@ -263,9 +263,19 @@ _Pointer _Refancy(_Pointer _Ptr) noexcept {
 }
 
 // FUNCTION TEMPLATE _Destroy_in_place
+template <class _NoThrowFwdIt, class _NoThrowSentinel>
+/* _CONSTEXPR20_DYNALLOC */ void _Destroy_range(_NoThrowFwdIt _First, const _NoThrowSentinel _Last) noexcept;
```
Comment (StephanTLavavej) [inline stl/inc/xmemory:267]: Drop top-level `const` from declaration:
```suggestion
/* _CONSTEXPR20_DYNALLOC */ void _Destroy_range(_NoThrowFwdIt _First, _NoThrowSentinel _Last) noexcept;
```

File: stl/inc/memory [hunk 2]
```diff
@@ -97,6 +97,62 @@ _NoThrowFwdIt uninitialized_move(const _InIt _First, const _InIt _Last, _NoThrow
...
+            _Seek_wrapped(_First1, _STD move(_UResult.in));
+            return {_STD move(_First1), _Rewrap_iterator(_Range2, _STD move(_UResult.out))};
+        }
+    };
+
+    inline constexpr _Uninitialized_move_fn uninitialized_move{_Not_quite_object::_Construct_tag{}};
+} // namespace ranges
+#endif // __cpp_lib_concepts
+
 // FUNCTION TEMPLATE uninitialized_move_n
```
Comment (StephanTLavavej) [inline stl/inc/memory:156]: Should `uninitialized_move_n` gain these `/* _CONSTEXPR20_DYNALLOC */` comments, or is a todo sufficient?

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+#include <algorithm>
+#include <cassert>
+#include <concepts>
+#include <memory>
+#include <ranges>
+#include <span>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:12]: Pre-existing aside: `P0896R4_ranges_subrange/test.cpp` alone uses double-quotes for `range_algorithm_support.hpp`.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+STATIC_ASSERT(same_as<decltype(ranges::uninitialized_move(borrowed<false>{}, borrowed<true>{})),
+    ranges::uninitialized_move_result<ranges::dangling, int*>>);
+STATIC_ASSERT(same_as<decltype(ranges::uninitialized_move(borrowed<true>{}, borrowed<false>{})),
+    ranges::uninitialized_move_result<int*, ranges::dangling>>);
+STATIC_ASSERT(same_as<decltype(ranges::uninitialized_move(borrowed<true>{}, borrowed<true>{})),
+    ranges::uninitialized_move_result<int*, int*>>);
+
+struct int_wrapper {
+    inline static int constructions = 0;
+    inline static int destructions  = 0;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:35]: Static data members can be inline?

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    static constexpr void clear_counts() {
+        if (!is_constant_evaluated()) {
+            constructions = destructions = 0;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:39]: I would prefer to see separate assignments, following our usual conventions.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    ranges::uninitialized_move_result<int*, ranges::dangling>>);
+STATIC_ASSERT(same_as<decltype(ranges::uninitialized_move(borrowed<true>{}, borrowed<true>{})),
+    ranges::uninitialized_move_result<int*, int*>>);
+
+struct int_wrapper {
+    inline static int constructions = 0;
+    inline static int destructions  = 0;
+
+    static constexpr void clear_counts() {
+        if (!is_constant_evaluated()) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:38]: Should include `<type_traits>`.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+};
+STATIC_ASSERT(movable<int_wrapper> && !copyable<int_wrapper>);
+
+template <class T, size_t N>
+struct holder {
+    T* ptr = allocator<T>{}.allocate(N);
+
+    holder()              = default;
+    holder(const holder&) = delete;
+    holder& operator=(const holder&) = delete;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:91]: Should implemented and deleted functions be separated by a newline?

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+template <test::ProxyRef IsProxy>
+using test_input  = test::range<test::input, int_wrapper, test::Sized::no, test::CanDifference::no, test::Common::no,
+    test::CanCompare::yes, IsProxy>;
+using test_output = test::range<test::fwd, int_wrapper, test::Sized::no, test::CanDifference::no, test::Common::no,
+    test::CanCompare::yes, test::ProxyRef::no>;
+
+constexpr void run_tests() {
+    // The algorithm is oblivious to non-required category, size, difference, and requires non-proxy references for the
+    // output range.
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:202]: This is hard to parse (I think because there's a list of "category, size, difference" without an "and", followed by a separate "and" clause).

File: stl/inc/memory (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/memory:156]: 1. I only annotated those functions that I actually tested in a constant expression during a temporary merge with @mnatsuhara's branch and another branch with `ranges::` flavors of `construct_at`, `destroy_at`, `destroy`, and `destroy_n`.
2. Double-checking the Standard, _none_ of the `uninit_meow` algorithms should have `constexpr`-related comments because they are not `constexpr`. Of the "specialized algorithms", only `construct_at`, `destroy_at`, `destroy`, and `destroy_n` (both `std` and `std::ranges` flavors, not including `ExecutionPolicy` overloads) are `constexpr`.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp (hunk 3 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:12]: Weird - I'll clean it up in this PR.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp (hunk 4 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:35]: Yes. `static constexpr` data members are implicitly `inline`, non-`constexpr` are not.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp (hunk 6 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:38]: No need now that there's nothing `constexpr`-related left in this test.

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp (hunk 7 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:91]: Yes, but mooted by simplifying this down to a non-`constexpr` wrapper around aligned storage for `N` `T`s.

File: stl/inc/memory [hunk 9]
```diff
@@ -97,6 +97,62 @@ _NoThrowFwdIt uninitialized_move(const _InIt _First, const _InIt _Last, _NoThrow
...
+
+    // VARIABLE ranges::uninitialized_move
+    class _Uninitialized_move_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_iterator _It1, sentinel_for<_It1> _Se1, _No_throw_forward_iterator _It2, _No_throw_sentinel_for<_It2> _Se2>
+            requires constructible_from<iter_value_t<_It2>, iter_rvalue_reference_t<_It1>>
+        /* _CONSTEXPR20_DYNALLOC */ uninitialized_move_result<_It1, _It2> operator()(
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/memory]: Yes: it's weird in this case because I've declared `ranges::_Uninitialized_move_unchecked` - which also returns `uninitialized_move_result` - in `<xmemory>` with the expectation that we might use it to implement `constexpr` containers. I'm now convinced that was a mistake, and that I should have put it here; we can easily promote things later.

File: stl/inc/xmemory [hunk 10]
```diff
@@ -1480,6 +1492,81 @@ _NoThrowFwdIt _Uninitialized_move_unchecked(_InIt _First, const _InIt _Last, _No
...
+
+    // clang-format off
+    // CONCEPT _No_throw_input_iterator
+    template <class _It>
+    concept _No_throw_input_iterator = input_iterator<_It>
+        && is_lvalue_reference_v<iter_reference_t<_It>>
+        && same_as<remove_cvref_t<iter_reference_t<_It>>, iter_value_t<_It>>;
+
+    // CONCEPT _No_throw_sentinel_for
+    template <class _Se, class _It>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xmemory:1535]: Nope: http://eel.is/c++draft/special.mem.concepts#4. _`no-throw-sentinel-for`_ vs. `sentinel_for` is a purely semantic distinction.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/xmemory:1535]: (Also concepts can't be constrained but they can refine other concepts; it would be `concept _No_throw_sentinel_for = _No_throw_input_iterator<_It> && sentinel_for<_Se, _It>;`.

**Comment by CaseyCarter** [comment]
There were substantial changes here after review, so I'm moving this back to Initial Review.

File: stl/inc/xmemory [hunk 11]
```diff
@@ -1480,6 +1492,40 @@ _NoThrowFwdIt _Uninitialized_move_unchecked(_InIt _First, const _InIt _Last, _No
 }
 #endif // _HAS_IF_CONSTEXPR
 
+#ifdef __cpp_lib_concepts
```
Comment (CaseyCarter) [inline stl/inc/xmemory:1495]: Is this the right place in this header?

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    STATIC_ASSERT(N < ~size_t{0} / sizeof(T));
+    alignas(T) unsigned char space[N * sizeof(T)];
+
+    auto as_span() {
+        return span<T, N>{reinterpret_cast<T*>(space + 0), N};
+    }
+};
+
+template <class R>
+void not_ranges_destroy(R&& r) {
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:84]:
```suggestion
void not_ranges_destroy(R&& r) { // TRANSITION, ranges::destroy
```

File: tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+template <test::ProxyRef IsProxy>
+using test_input  = test::range<test::input, int_wrapper, test::Sized::no, test::CanDifference::no, test::Common::no,
+    test::CanCompare::yes, IsProxy>;
+using test_output = test::range<test::fwd, int_wrapper, test::Sized::no, test::CanDifference::no, test::Common::no,
+    test::CanCompare::yes, test::ProxyRef::no>;
+
+int main() {
+    // The algorithm is oblivious to non-required category, size, difference, and "proxyness" of the input range. It
+    // _is_ sensitive to proxiness in that it requires non-proxy references for the output range.
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_uninitialized_move/test.cpp:175]: proxyness versus proxiness.

File: stl/inc/memory [hunk 14]
```diff
@@ -97,6 +130,73 @@ _NoThrowFwdIt uninitialized_move(const _InIt _First, const _InIt _Last, _NoThrow
...
+
+            _Uninitialized_backout _Backout{_STD move(_OFirst)};
+
+            for (; _IFirst != _ILast && _Backout._Last != _OLast; ++_IFirst) {
+                _Backout._Emplace_back(_RANGES iter_move(_IFirst));
+            }
+
+            _OFirst = _STD move(_Backout._Last);
+            _Backout._Release();
+            return {_STD move(_IFirst), _STD move(_OFirst)};
```
Comment (miscco) [inline stl/inc/memory:192]: Out of curiosity, is that dance really needed?

Could we simply do:
```suggestion
            return {_STD move(_IFirst), _STD move(_Backout._Release())};

```

File: tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp [hunk 15]
```diff
@@ -983,3 +984,128 @@ namespace gh_1089 {
...
+    STATIC_ASSERT(same_reference_value<iterator_archetype<iterator_status::forward>>);
+
+    // Validate _No_throw_input_iterator
+    STATIC_ASSERT(!_No_throw_input_iterator<iterator_archetype<iterator_status::not_input>>);
+    STATIC_ASSERT(!_No_throw_input_iterator<iterator_archetype<iterator_status::not_lvalue_reference>>);
+    STATIC_ASSERT(!_No_throw_input_iterator<iterator_archetype<iterator_status::different_reference_and_value>>);
+    STATIC_ASSERT(_No_throw_input_iterator<iterator_archetype<iterator_status::input>>);
+    STATIC_ASSERT(_No_throw_input_iterator<iterator_archetype<iterator_status::forward>>);
+
+    enum class sentinel_status : int { not_sentinel, good };
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp:1048]: I have a mild preference for symmetric(ish) naming here: `{not_sentinel, is_sentinel}` or `{bad, good}`.  Perhaps even `{no, yes}` as I think that would read nicely ("`sentinel_status::no`", "`sentinel_status::yes`").  
Feel free to ignore.

File: tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp [hunk 16]
```diff
@@ -983,3 +984,128 @@ namespace gh_1089 {
...
+    STATIC_ASSERT(ranges::range<range_archetype<range_status::input>>);
+    STATIC_ASSERT(ranges::range<range_archetype<range_status::forward>>);
+
+    // Validate _No_throw_input_range; note that the distinction betweeen range<R> and
+    // no-throw-sentinel-for<sentinel_t<R>, iterator_t<R>> is purely semantic, so we can't test them separately.
+    STATIC_ASSERT(!_No_throw_input_range<range_archetype<range_status::not_range>>);
+    STATIC_ASSERT(!_No_throw_input_range<range_archetype<range_status::not_input>>);
+    STATIC_ASSERT(_No_throw_input_range<range_archetype<range_status::input>>);
+    STATIC_ASSERT(_No_throw_input_range<range_archetype<range_status::forward>>);
+
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp:1106]:
```suggestion

    // Validate _No_throw_forward_range
```
for consistency with the rest of the `STATIC_ASSERT` blocks.

File: stl/inc/memory (hunk 14 above)
Comment (CaseyCarter) [reply stl/inc/memory:192]: I've been rewrapping iterators for too long to stop now! (Yes, this is a good change.)

File: stl/inc/memory [hunk 17]
```diff
@@ -97,6 +130,73 @@ _NoThrowFwdIt uninitialized_move(const _InIt _First, const _InIt _Last, _NoThrow
...
+            auto _UResult = _Uninitialized_move_unchecked(
+                _Get_unwrapped(_STD move(_First1)), _Uend(_Range1), _Ubegin(_Range2), _Uend(_Range2));
+
+            _Seek_wrapped(_First1, _STD move(_UResult.in));
+            return {_STD move(_First1), _Rewrap_iterator(_Range2, _STD move(_UResult.out))};
+        }
+
+    private:
+        template <class _It1, class _Se1, class _It2, class _Se2>
+        _NODISCARD static uninitialized_move_result<_It1, _It2> _Uninitialized_move_unchecked(
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/memory:176]: The convention is: `_Meow_unchecked` takes unwrapped iterator(s) and sentinel(s) of possibly differing types and `meows` them, whereas `_Meow_common` takes unwrapped iterator(s) and sentinel(s) of the same type (what we called "iterator pairs" before C++20). This function allows differing iterator and sentinel types so it should be `_unchecked`.

The convention change in GH-1127 was that names should now always follow the rules above, whereas previously I was only using `std::ranges::_Meow_common` if there was a corresponding `std::_Meow_unchecked` that I wanted to clearly differentiate from, and would otherwise just default to `std::ranges::_Meow_unchecked` whether or not the function required identical iterator and sentinel types.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/memory:176]: TLDR: No it should not ;)

File: tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp (hunk 16 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp:1106]: Weird, I must have ran out of `/`s by the time I got here. Fixed.

File: tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp (hunk 15 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_algorithm_machinery/test.cpp:1048]: `{ no, yes }` is fine here.

**Comment by StephanTLavavej** [comment]
Thanks! I don't have a funny joke here - this is simply raw space where a funny joke can be move-constructed later.
